README.txt
BMF, the bi-directional machining facility is a PC-based interface
for embedded controllers.  the main interface, obmf.py, communicates
with an embedded system over a serial interface.  

The protocol used for data exchange is described in detail in PROTOCOL.txt
In general, it consists of OpCodes (single byte commands) followed by the 
arguments for each command, and lines usually terminated by a line feed 
character.  The exception to this is that intel-hex format is used to 
transfer file data.  The protocol is symmetrical in that either the PC or 
the embedded host can send commands.


Requirements
------------

You need a PC (windows or Linux), running python ( 2.5 < x < 3 ) and QT
bindings (pyQT4, pyside not tested yet.), and an embedded processor, on which
you have implemented the protocol, and a serial interface connection between 

The serial interface can be either via a hardware serial port, a port 
connected via a USB-serial cable, or even connection via a terminal-server 
over a network connection, wired or wireless.   The application should 
auto-discover all local serial ports, but remote ports would have to be
entered manually.

Startup
-------

open a command window, navigate to the directory containing obmf.py,
then type:

python ./obmf.py



Testing Without a Device
------------------------

To try out the software without a device that implements the protocol, there 
is an emulation mode which replaces the actual device by a software 
implementation.  One can start up the tool in a embedded processor emulation 
mode, like so::

	python obmf.py --port="localhost:5050" --flags=4 read

which reads the commands sent to it by a client connecting to the
given port.  A client is then started up in a conventional way::

        python obmf.py --port="localhost:5050" --flags=8 view

You can do the same with a nullmodem cable:
	python obmf.py --port=/dev/ttyUSB1 --speed=115200 read
	python obmf.py --port=/dev/ttyUSB0 --speed=115200 view

Under View->Testing, one can bring up the arbitrary hex code dialog, and
then send the magic sequence AA,A (0xAA,0xA) which will trigger code in
the emulator to send many commands, putting strings on the screen, updating
counters, and setting key labels.   This demonstrates the results which
the embedded processor is able to perform.





CHANGES.txt

WORKING ON:
  -- add mouseevents in char display.
  -- bug in font dialog causes system crash (Linux only.)

TODO:

  -- five digits with a . and all 0's. pain in the butt. code size already tripled, with no leading digits.
  -- logfile support not general yet.
  -- declarative UI? or CSS or ...
  -- Replace literal op-codes with consants OC_
  -- set cwd ?
  -- decimal in counters.
  -- documentation review.
        -- help function in GUI?
  -- automated Testing.
        -- file transfer loopback test.

  -- logging to a text file... 
        -- add file name choice to Settings.
        -- good method of logfile name selection.
        -- maybe z80 will log too?
  -- have routine updates make radiobuttons symmetrical.
  -- have toggles stay depressed when appropriiate.

2011/09/10:
  -- have startup work properly... using command line or reading from saved state.
  -- corrected vertical spacing in character display by using actual font height.

2011/02/20:
  -- Font Selection dialog in Character mapped display
  -- GUI save/restore bug: doesn't seem to save/restore the size of the main window?


2011/01/23:
  -- implemented save/restore state (GUI and comm parameters)
  -- save keyboard positions and settings.
  -- Align Right labels...
  -- fixed bug: click on flags causes value change...

december:
  -- implement new opcode 82.
  -- default serial baudrate now 115200 (embedded system updated.)
  -- options for flow control (still off) included in code.
  -- removed 0a from counter update opcode.
  -- simplified key event processing (single routine replaces two)


2010/11/28:
  -- pad with only NULS.
  -- removed h/w flow control (added before undocumented.)

2010/11/27:
  -- added "natural" flow control, requires an ack for each key press,
     discards any keypresses received when last one not acked.
  -- Added Trace function.
  -- added return binding for faster hex code send testing.
  -- bound 2f key to sendfile function (functionality lost in panel reset...)
  -- separate self-test function from normal stuff.

2010/11/11:
  -- bugfixes to label setting, added utf-8 support for strings.

2010/11/06:
  -- 2x 16 key pads, programmed with set string op-code.

2010/10/17:
  -- bug-fixes to file sending (name handling was mangling them.)
  -- added leds and labels to the loopback tests.
  -- determined that neither worked, bugfixed.

2010/10/12: 
  -- pullfile function.
      -- complete readcmd stanza for 0x87
  -- Added file name to 0x80 blockmode, becomes a send file op-code.
  -- FIXME: reception of intel hex format.
	-- vrfy checksums.
	-- store them to file, named by 80 op-code 
  -- verification receipt of intel-hex records. Re-Tx requested as needed.
  -- removed display switches in Settings Menu, redundant vs. View menu.
  -- re-tested file transfer, might have been busted, misc. bug-fixes. works now.
  -- logging to a text file... 
        -- log everything...
  -- make leds work?
        -- should clicking cause an action?
        -- status update via flags.
        -- label set via op-code.

2010/10/06:
  -- made keyboard faster: eliminate long autorepeat delay, shorten interval 10x to 5 ms.
  -- add LEDS (visual only for now...)
  -- changed text buttons to arrows.
  -- go routines removed from GUI, and a single go routine in __init__.py
  -- go routine now returns a counter update to BMF peer (emulation function.)
  -- go routines now called on receipt of key stroke.
  -- eliminated last_update
  -- eliminate lots of dead code:
       -- self.last_update in GUI (replaced by timed stuff.)

2010/10/05:
  -- added leading 0´s to character display of counter values.
  -- Docking Widget re-factor.

2010/10/02:
  -- documentated purpose for many routines in GUI.py.
  -- added 0x90->0x9f op-codes for counter updates.
  -- added 0x84 setCounterXY support, new op-code to say where it lives.
  -- have sub-second timestamps in the log.
  -- replaced fixed refresh interval with looking at how long refresh takes
     on the machine, and trying not to return before it has spent at least
     half it's time on something else.  In this way, it should never lag... 
  -- made protocol symmetrical
       -- tx register setting. (one liner!)
       -- tx graphic display  (one liner!)
       -- tx register positioning (one liner!)...
  -- Added the AA - Alcaholics Anonymous op-code...
     improved self-test rig, by having network server generate test data, so 
     GUI actually receives and parses display update commands (this code not
     exercised in previous testing rigs.)

2010/09/25:
  -- speedups:
	-- optimized some loops. (CharDisplay.repaint...)
	-- ignore graphical updates < x ms. apart (see GUI.__guiupdate)
        -- started buffering reads
  -- added crash and burn to avoid infinite loop at exit of one end of socket.
  -- from: make protocol symmetrical
       -- rx intel hex stream... (bogus, just discards what it rx's)
  -- log length now fixed at 100.

2010/09/21:
  -- bugfix regressions for intel hex send (binrec)
  -- bugfix regression for send button binding.
  -- bugfix clear screen decode meant nothing gets displayed.
  -- enhancement: hello prints in various places. 

2010/09/19:
  -- have font selection fill window in char display
  -- disconnect.
  -- clear screen command in Testing.
  -- remove autorepeat from Connect.

2010/09/18:
  -- have toggle buttons visually confirm state.
  -- make protocol symmetrical
       -- rx register setting
       -- rx draw command.
  -- use poll & serial peek commands to remove non-blocking reads.
  -- starts up with no connection (need not have a bmf object working.)
  -- move logging to work with callbacks, so other classes can log.
  -- have protocol for drawing in the char window.
  -- have it work with sockets, and have two bmf's talk to eachother.
  -- get on github. need public ssh key for that.
  -- send multiple arbitrary bytes (>1)
  -- have Settings dialog work.


2010/09/10:
----------

obmf.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Copyright (C) 2008,2009  Peter Silva (Peter.A.Silva@gmail.com)
comes with ABSOLUTELY NO WARRANTY; For details type see the file
named COPYING in the root of the source directory tree.
"""

"""
Operate BMF - wrapper to operate the bi-directional machining facility.

 usually used to fire up a GUI to talk to the BMF over a serial line, that would just be:

      obmf.py

 (default is ´view´ to fire up the GUI, and one can use the settings tab to identify the port.)

      obmf.py --port=localhost:50007 --flags=4 read
         - start up a line mode interface that will only read commands, in network server mode.


      obmf.py --port=localhost:5007 --flasgs=8 view 
         - start up a gui in net client mode, will connect to above...


"""

import bmf
import string
import time
import platform

def usage():
  print """
  obmf <options> command <arguments> 
    -p / --port = "COM2:", or "/dev/ttyS0"
    -s / --speed = 38400 
    -l / --logfile = "bmf.log"
    -f / --flags = <flags>    
           flags:  1 - simulate: port is a write only file.
		   2 - require keys sent to be acknowledged.
                   4 - network server
		   8 - network client
                  16 - print TRACE output 
                       (hexlification of every byte transferred)
    -h / --help  -- print this usage information.

  command is one of:
        view  - start the keyboard GUI.
        send file.hex  - write a file in intel hex format 0
             file.bin  - write a binary, encoding in intel hex.
 
        key  <command>
        key is one of: """ + string.join(bmf.keycodes.keys(),',') + """
  """


import sys
import getopt

last_update=0

def print_msg(s):

    global last_update

    now=time.time()
    print '%s - %s' % ( str(time.time()), s)


def operate_bmf(port=None,cmd="view",speed=38400,dbg=0):
    
    #if port != None:
    b = bmf.bmf(dev=port,speed=speed,flags=dbg,msgcallback=print_msg)
    #else:
    #  b = None
    
    if cmd == 'view':
      from PyQt4 import QtGui
      from PyQt4 import QtCore
      from bmf.GUI import GUI
     
      app = QtGui.QApplication(sys.argv)
      #app.setStyle("windows")
      #app.setStyle("macintosh")
      #app.setStyle("plastique")
      #app.setStyle("windowsxp")
      tb = GUI(b)
      tb.show()
      QtGui.qApp = app
      app.exec_()
      QtGui.qApp = None
      app = None
      return
 
    b.connect()
    if cmd  == 'key' :
      print "Send a key"
      b.sendKey(args[1])
    elif cmd == 'read' :
      while True:
          b.readcmd(True)
    elif cmd == 'test':
          b.sendTestCommands()
    elif cmd == 'send':
      print "Send Intel Hex file"
      filename = args[1]
      if (filename[-4:].lower() == '.hex'):
         b.sendbulkhex(filename)
      else:
         b.sendbulkbin(filename)
    
    else:
      print "some other command..."

if __name__ == '__main__':
    #port="COM2:"

    #FIXME: code for testing purposes...
    #    hardcoded defaults to make testing go quickly.
    #    would be nice to replace with a detection scan...
    #
    #if platform.system() == "Windows":
    #    port="COM1"
    #    dbg=0
    #else:
    #    # I test with a socket back to myself now...
    #    port='localhost:50007'
    #    dbg=8

    speed=115200
    port=None 
    dbg=0

    opts, args = getopt.getopt(sys.argv[1:],"f:hp:s:V",[ 
    "flags=", "f=", "help", "port=", "speed=", "version" ])
    
    for o, a in opts:
      if o in ( "-f", "--f", "--flags"):
         dbg = int(a)
      elif o in ( "-l", "--logfile"):
         logfile = a
      elif o in ( "-p", "--port"):
         port = a
      elif o in ( "-s", "--speed"):
         speed = int(a)
      elif o in ( "-V", "--version"):
         print "my version..."
      else:  # includes -h --help...
         usage()
         sys.exit()

    if len(args) == 0:
      cmd = 'view'
    else:
      cmd = args[0]

    operate_bmf(port,cmd,speed,dbg)
